dbms/src/Storages/Page/V3/WALStore.cpp

```diff
@@ -270,4 +278,48 @@ bool WALStore::saveSnapshot(
     return true;
 }
 
+void WALStore::removeLogFiles(const LogFilenameSet & log_filenames)
+{
+    for (const auto & filename : log_filenames)
+    {
+        const auto log_fullname = filename.fullname(LogFileStage::Normal);
+        provider->deleteRegularFile(log_fullname, EncryptionPath(log_fullname, ""));
+    }
+    {
+        std::unique_lock<std::mutex> lock(log_file_max_sequences_cache_mutex);
+        for (const auto & filename : log_filenames)
+        {
+            log_file_max_sequences_cache.erase(filename);
+        }
+    }
+}
+
+UInt64 WALStore::getLogFileMaxSequence(const LogFilename & log_filename, std::function<UInt64(const String & record)> max_sequence_getter)
+{
+    std::unique_lock<std::mutex> lock(log_file_max_sequences_cache_mutex);
+    auto iter = log_file_max_sequences_cache.find(log_filename);
+    if (iter != log_file_max_sequences_cache.end())
+        return iter->second;
+    else
+    {
+        auto log_reader = wal_store_reader->createLogReader(log_filename);
```

Comment:
Can we make `WALStoreReader::createLogReader` to be a static method? It is weird to me that we need to keep a `wal_store_reader` as long as `WALStore`

Reply:
Fixed.